qa/rpc-tests/randompruning.py

```diff
@@ -0,0 +1,101 @@
+        # pick this one to start from the cached 4 node 100 blocks mined configuration
+        # initialize_chain(self.options.tmpdir, bitcoinConfDict, wallets)
+        # pick this one to start at 0 mined blocks
+        initialize_chain_clean(self.options.tmpdir, 4, bitcoinConfDict, wallets)
+        # Number of nodes to initialize ----------> ^
+
+    def setup_network(self, split=False):
+        self.nodes = []
+        self.nodes.append(start_node(0, self.options.tmpdir, ["-debug", "-useblockdb=1"]))
+        self.nodes.append(start_node(1, self.options.tmpdir, ["-debug=prune", "-useblockdb=1", "-prunewithmask=1"]))
+
+        # Now interconnect the nodes
+        connect_nodes_full(self.nodes)
+        # Let the framework know if the network is fully connected.
+        # If not, the framework assumes this partition: (0,1) and (2,3)
+        # For more complex partitions, you can't use the self.sync* member functions
+        self.is_network_split=False
+        self.sync_blocks()
+
+    def run_test (self):
+        # generate enough blocks so that nodes[0] has a balance
+        # mine some number of blocks higher than the minimum we would be pruning
+        self.sync_blocks()
+        blocks_mined = []
+        for i in range(101):
+            blockhash = self.nodes[0].generate(1)[0]
+            # dont add blocks above the min blocks to keep line
+            if (i < 80):
+                blocks_mined.append(blockhash)
+        self.sync_blocks()
+
+        #get the max possible value of our hashmask
+        uint64_t_max = (2**64) - 1
+        # calculate 1% which
+        threshold_percent = uint64_t_max / 100
+        blockchaininfo = self.nodes[1].getblockchaininfo()
+        fullhashMask = blockchaininfo["pruneHashMask"]
+        threshold = blockchaininfo["hashMaskThreshold"]
+        # the mask is stored as a 256 bit int for easy bitwise operations with block hashes
+        # get the 64 LSB (everything else should just be 0's)
+        hashMask64 = GetLow64(fullhashMask)
+        # calculate the normalize threshold for the node
+        normalized_threshold = threshold * threshold_percent
+
+        # check the blocks mined prior to determine if we are keeping the blocks we expect to have
+        # and pruning the ones we expect to prune
+
+        # kept does nothing in the actual test but it was/is useful for debugging if the test fails
+        kept = []
+        for block in blocks_mined:
+            assert_equal(len(block), 64)
+            low64block = GetLow64(block)
+            # if the 64LSB of the blockhash is equal to or above our threshold we should have pruned it
+            # assert this is the case
+            if ((int(low64block, 16) ^ int(hashMask64, 16)) >= normalized_threshold):
+                assert_raises_rpc_error(-32603, "Block not available (pruned data)", self.nodes[1].getblock, block)
+            else:
+                kept.append(self.nodes[1].getblock(block)['height'])
+
```

Comment:
shouldn't we also check that we kept the blocks we expected to?

Reply:
doesnt this already do that? if we dont raise the assert where we think we should, then it implies that are keeping everything we expect to keep?